### ISSUE TITLE
Improve data extraction in `dscquery` to avoid loading the same column of files multiple times

### ISSUE THREAD

**Issue by gaow** [stephenslab/dsc#168]
I've got a benchmark result with a meta-data table dimension ~7,000 by ~10. The query looks like:

```r
> out = dscquery('finemap_output', "small_data.dataset sharing_pattern.n_signal mnm.eff_mode susie_scores.total susie_scores.valid susie_scores.size susie_scores.purity susie_scores.top")
Loading dsc-query output from CSV file.
Reading DSC outputs:
 - sharing_pattern.n_signal: extracted atomic values
 - susie_scores.total: extracted atomic values
 - susie_scores.valid: extracted atomic values
 - susie_scores.size: extracted atomic values
 - susie_scores.purity: extracted atomic values
 - susie_scores.top: extracted atomic values
```

You see 6 columns have been extracted. Using a single core `lapply` it took 3.511436 mins to extract it all. 

There are two issues to discuss:

1. Should we revert and improve on the option to load data in parallel?
2. Currently I believe we load `susie_scores` file 5 times for the different 5 values requested. A more obvious improvement is to create hierarchies of files then values under files so that we only load each file at most once.

@pcarbo we can table 1 as discussed on slack, but I wonder if you'd like to help implementing 2.

**Comment by pcarbo** [comment]
@gaow Read this before we proceeding, and let me know your thoughts: https://stackoverflow.com/questions/5321768/how-many-threads-for-reading-and-writing-to-the-hard-disk

**Comment by gaow** [comment]
Thanks @pcarbo. As proposed above I agree to not do 1 at the moment. However the discussion above is limited to a single I/O queue on a conventional HDD. It did not discuss performance on SSD or file systems like RCC. Even so, it was not clear if the I/O process they consider users all the I/O bandwidth (which is not our case because we deal with many tiny RDS files). 

I'm more concerned with 2 at this point. We can possibly save 2/3 total I/O computations if we can avoid loading the same file multiple times.

**Comment by pcarbo** [comment]
> I'm more concerned with 2 at this point. We can possibly save 2/3 total I/O computations if we can avoid loading the same file multiple times.

Yes! Could you put together a fully runnable example I can use as a test case? Then I will try to implement this in `dscquery`.

**Comment by gaow** [comment]
@pcarbo certainly, thanks a lot! I'll try to get to it by the end of today after wrapping up something else now.

**Comment by gaow** [comment]
@pcarbo here is a toy example.

DSC file:

```
simulate: R()
  $out1: 1
  $out2: 2

method: R()
  out1: $out1
  out2: $out2
  $out3: 3
  $out4: 4

DSC:
  run: simulate * method
```

to run it:

```
dsc test.dsc --replicate 10
```

to query and extract:

```r
> library(dscrutils)
> dscquery('test', 'simulate.out1 simulate.out2 method.out3 method.out4')
INFO: Loading database ...
INFO: Running queries ...
INFO: Extraction complete!
Loading dsc-query output from CSV file.
Reading DSC outputs:
 - simulate.out1: extracted atomic values
 - simulate.out2: extracted atomic values
 - method.out3: extracted atomic values
 - method.out4: extracted atomic values
   DSC simulate.out1 simulate.out2 method.out3 method.out4
1    1             1             2           3           4
2    2             1             2           3           4
3    3             1             2           3           4
4    4             1             2           3           4
5    5             1             2           3           4
6    6             1             2           3           4
7    7             1             2           3           4
8    8             1             2           3           4
9    9             1             2           3           4
10  10             1             2           3           4
> 
```

Thanks!

**Comment by pcarbo** [comment]
Another slightly more complicated example.

```yaml
simulate: R()
  $out1: 1
  $out2: 2

method1: R()
  out1: $out1
  out2: $out2
  $out3: 3
  $out4: 4

method2: R()
  out1: $out1
  out2: $out2
  $out3: 3
  $out4: 4

DSC:
  define:
    method: method1, method2
  run: simulate * method
```

**Comment by pcarbo** [comment]
@gaow This should be more efficient now. All tests pass. See commit `7c5ea6b`. But it would be good if you could compare results before & after on a more complicated example in which the results differ among replicates and modules. Also, it would be good if you could spin up an example that has missing (NA) results files, since the logic is fairly complicated for handling this (I keep track of this with the `available.targets` variable inside function `dscquery`).

Let me know if you have any questions about the implementation.

**Comment by gaow** [comment]
@pcarbo sorry for the delayed reply ... Thanks for the work! I tried it out. Unfortunately there was no improvement on a non-trivial example I had compared to previous version: the extraction took 3.5 minutes. I was reading the patch -- it is not obvious to me that all files are loaded at most once for all quantities to extract. Maybe I misread it, but thought it'd be easier to verify it here with you that this is the case. If so, then I guess there is little we can improve? 3.5 minutes is not seriously wrong, but not good either, for a result matrix of 7350 by 14.

**Comment by pcarbo** [comment]
@gaow One way to convince yourself (or to prove me wrong!) is to add this to the first line of `read_dsc` in `utils.R`:

```R
cat("*\n")
```

If I implemented this correctly, the number of *'s in the above example should be fewer after my change.

**Comment by gaow** [comment]
@pcarbo to sum up our discussion what you've implemented previously will make per column extraction faster when there are multiple duplicates in the column (result of a module shared by multiple downstream modules). However we are more worried about columns themselves being duplicate of one another, as with the case when people want to extract multiple quantities from the same module, say different scores from a score module. Those columns will still get to load multiple times. 

To consolidates duplicated columns when it comes to loading data, we need to change the structure of the code a lot because currently the for loop is over columns.
